rosidl_parser/rosidl_parser/grammar.lark

```diff
@@ -75,13 +76,7 @@ string_literal: "\"\"" | "\"" /(\\\"|[^"])+/ "\""
 wide_string_literal: "L\"\"" | "L\"" /(\\\"|[^"])+/ "\""
 
 // 7.2.6.4 Floating-point Literals
-floating_pt_literal: DIGIT+ floating_pt_literal_dot DIGIT+ floating_pt_literal_e DIGIT*
-  | floating_pt_literal_dot DIGIT+ (floating_pt_literal_e DIGIT*)?
-  | DIGIT+ floating_pt_literal_dot DIGIT* (floating_pt_literal_e DIGIT*)?
-  | DIGIT+ floating_pt_literal_e DIGIT*
-// separate rules to identify the components
-floating_pt_literal_dot: "."
-floating_pt_literal_e: "e"i
+floating_pt_literal: FLOAT
```

Comment:
`FLOAT` is defined as follows:

> `FLOAT: INT _EXP | DECIMAL _EXP?`

and the latter part doesn't match the definition of `floating_pt_literal` in the IDL spec. Therefore I don't think it can be used as is.

Reply:
Hm, I'm not sure I agree.  7.2.6.4 in the IDL spec says:

```
A floating-point literal consists of an integer part, a decimal point (.), a fraction part, an e or E, and an optionally
signed integer exponent. The integer and fraction parts both consist of a sequence of decimal (base ten) digits. Either
the integer part or the fraction part (but not both) may be missing; either the decimal point or the letter e (or E) and the
exponent (but not both) may be missing.
```

That seems to say to me that the following cases are allowed (everything else is denied):
* "1.1e10"
* "1.1e-10"
* "1.1"
* "1."
* ".1"
* "1e-5"
* "1e5"

Looking through the grammar for the Lark built-in FLOAT, it looks like:

```
INT: DIGIT+
SIGNED_INT: ["+"|"-"] INT
DECIMAL: INT "." INT? | "." INT

// float = /-?\d+(\.\d+)?([eE][+-]?\d+)?/
_EXP: ("e"|"E") SIGNED_INT
FLOAT: INT _EXP | DECIMAL _EXP?
```

That seems to allow the same cases.  However, I might be missing something, so could you explain a case that doesn't work with that grammar, but is allowed by IDL (or vice-versa)?

Reply:
`FLOAT` allows a simple `DECIMAL` which in the IDL spec is not a `floating_pt_literal` but a `fixed_pt_literal`.

Reply:
But `7.2.6.5` says:

```
A fixed-point decimal literal consists of an integer part, a decimal point (.) , a fraction part and a d or D. The integer and fraction parts both consist of a sequence of decimal (base 10) digits. Either the integer part or the fraction part (but not both) may be missing; the decimal point (but not the letter d or D) may be missing.
```

So my understanding is that any fixed point has to end with a literal `d` or `D`, which is not the case for FLOAT.

Reply:
You are right that 7.2.6.5 / `fixed_pt_literal` is different since it must have a trailing `d` / `D`.

My previous point about 7.2.6.4 still stands: 

> `FLOAT` allows a simple `DECIMAL` which in the IDL spec is not a `floating_pt_literal` (7.2.6.4)

7.2.6.4 says:

> ... either the decimal point or the letter e (or E ) and the
exponent (**but not both**) may be missing.

Reply:
> `FLOAT` allows a simple `DECIMAL` which in the IDL spec is not a `floating_pt_literal` (7.2.6.4)

Just to be clear: DECIMAL is not the same thing as INT.  DECIMAL is defined in the lark grammar as:

```
DECIMAL: INT "." INT? | "." INT
```

That is, "5.", "5.0", or ".5" are what is allowed.  An integer such as "5" won't match a DECIMAL (though it would match an INT).

And since `FLOAT` is defined as:

```
FLOAT: INT _EXP | DECIMAL _EXP?
```

That takes care of having either an `e|E` or a `.`, or both, but not neither.  Thus, I still think that we can reuse FLOAT here.

If you still don't agree, can you give me an example of a number that you think is allowed by the IDL spec, but is not allowed for lark's FLOAT type?  Or one that is not allowed by the IDL spec, but is allowed for lark's FLOAT type?

Reply:
> And since `FLOAT` is defined as:
> `FLOAT: INT _EXP | DECIMAL _EXP?`
> That takes care of having either an `e|E` or a `.`, or both, but not neither.  Thus, I still think that we can reuse FLOAT here.

:+1:  I think I got it now and agree with you on: `floating_pt_literal: FLOAT`.